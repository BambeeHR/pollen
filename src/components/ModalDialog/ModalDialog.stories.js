// import { action } from '@storybook/addon-actions';
import { boolean, select, text } from '@storybook/addon-knobs';
import BaseButton from '../BaseButton/BaseButton.vue';
import ModalDialog from './ModalDialog.vue';
import Modal from '../../constants/Modal';

export default {
  title: 'ModalDialog',
  component: ModalDialog,
  parameters: {
    componentSubtitle: '',
  },
};

export const Default = () => ({
  components: { BaseButton, ModalDialog },
  props: {
    actionAlignment: {
      default: select(
        'Action Alignment',
        Object.values(Modal.ActionAlignments),
        Modal.ActionAlignments.CENTER
      ),
    },
    fullBleed: {
      default: boolean('Full Bleed', false),
    },
    header: {
      default: select(
        'Header',
        Object.values(Modal.Headers),
        Modal.Headers.TRANSPARENT
      ),
    },
    overlay: {
      default: select(
        'Overlay',
        Object.values(Modal.Overlays),
        Modal.Overlays.PRIMARY
      ),
    },
    overlayClose: {
      default: boolean('Close on overlay click', true),
    },
    size: {
      default: select('Size', Object.values(Modal.Sizes), Modal.Sizes.REGULAR),
    },
    subfooter: {
      default: text('Subfooter', 'Subfooter message goes here'),
    },
    showDelete: {
      default: boolean('Show delete', false),
    },
  },
  data() {
    return { isModalOpen: true };
  },
  computed: {
    actions() {
      return {
        primary: {
          label: 'Accept',
        },
        secondary: {
          label: 'Skip',
        },
        tertiary: this.showDelete
          ? {
              label: 'Delete',
              preIcon: 'trash',
            }
          : null,
      };
    },
  },
  template: `
  <div>
    <BaseButton @click="isModalOpen = !isModalOpen">Toggle Modal</BaseButton>
    <ModalDialog
      v-if="isModalOpen"
      :action-alignment="actionAlignment"
      :actions="actions"
      :full-bleed="fullBleed"
      :header="header"
      :overlay="overlay"
      :overlay-close="overlayClose"
      :size="size"
      @close="isModalOpen = false">
      <template #header>
        Heading Goes Here
      </template>
      Content
      <template v-if="subfooter" #subfooter>
        {{ subfooter }}
      </template>
    </ModalDialog>
  </div>
  `,
});
