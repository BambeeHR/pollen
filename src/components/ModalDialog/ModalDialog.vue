<template>
  <FocusLock class="modal-dialog">
    <ModalWindow :id="id" :size="size" :full-bleed="fullBleed">
      <ModalHeader
        v-if="!!$slots.header"
        :id="`${id}-header`"
        :variant="header"
      >
        <slot name="header" />
      </ModalHeader>
      <div :id="`${id}-body`" class="modal-dialog__body">
        <slot />
      </div>
      <ModalActions
        v-if="actions"
        :alignment="actionAlignment"
        :primary-action="actions.primary"
        :secondary-action="actions.secondary"
        :tertiary-action="actions.tertiary"
      />
      <ModalSubfooter v-if="!!$slots.subfooter" class="modal-dialog__subfooter">
        <slot name="subfooter" />
      </ModalSubfooter>
    </ModalWindow>
    <ModalClose :variant="overlay" @click="close" />
    <ModalOverlay :variant="overlay" @click="handleOverlayClick" />
  </FocusLock>
</template>

<script>
import FocusLock from 'vue-focus-lock';
import shortid from 'shortid';
import Modal from '../../constants/Modal';
import ModalActions from '../internal/modal/ModalActions.vue';
import ModalClose from '../internal/modal/ModalClose.vue';
import ModalHeader from '../internal/modal/ModalHeader.vue';
import ModalOverlay from '../internal/modal/ModalOverlay.vue';
import ModalSubfooter from '../internal/modal/ModalSubfooter.vue';
import ModalWindow from '../internal/modal/ModalWindow.vue';

/**
 * import { ModalDialog } from '@bambee/pollen';
 *
 * A modal dialog layout component. You'll likely want to use something like
 * vue-portals to properly place this component within the larger page layout.
 * It's also up to the parent component's statefulness to determine visibility
 * via a `v-if` conditional.
 */
export default {
  components: {
    FocusLock,
    ModalActions,
    ModalClose,
    ModalHeader,
    ModalOverlay,
    ModalSubfooter,
    ModalWindow,
  },
  props: {
    /**
     * Alignment of the action buttons. One of `center`, `right`, and `justify`.
     */
    actionAlignment: {
      type: String,
      default: Modal.ActionAlignments.CENTER,
      validator: (value) =>
        Object.values(Modal.ActionAlignments).includes(value),
    },
    /**
     * An object with optional `primary`, `secondary`, and `tertiary` keys.
     * These each should have at the very least a `label` string and `handler`
     * function. Any additional values will be spread on to an underlying
     * `BaseButton`.
     */
    actions: {
      type: Object,
      default: null,
    },
    /** If true, removes the padding within the modal window. */
    fullBleed: {
      type: Boolean,
      default: false,
    },
    /**
     * The header style. One of `transparent`, `primary`, `secondary`,
     * `warning`, `error`, `primary-ornate`, `warning-ornate`, or `error-ornate`
     */
    header: {
      type: String,
      default: Modal.Headers.TRANSPARENT,
      validator: (value) => Object.values(Modal.Headers).includes(value),
    },
    /** A unique element ID. By default, one is randomly generated. */
    id: {
      type: String,
      default: shortid.generate,
    },
    /**
     * The background overlay style. One of `primary`, `dark`, or `light`.
     */
    overlay: {
      type: String,
      default: Modal.Overlays.PRIMARY,
      validator: (value) => Object.values(Modal.Overlays).includes(value),
    },
    /** If true, emits `close` event when the background overlay is clicked. */
    overlayClose: {
      type: Boolean,
      default: true,
    },
    /**
     * The overlay width. One of `tiny`, `small`, `regular`, `medium`, `large,
     * or `giant`.
     */
    size: {
      type: String,
      default: Modal.Sizes.REGULAR,
      validator: (value) => Object.values(Modal.Sizes).includes(value),
    },
  },
  mounted() {
    document.addEventListener('keyup', this.handleKeyup);
    document.body.classList.add('overflow-hidden');
  },

  beforeDestroy() {
    document.removeEventListener('keyup', this.handleKeyup);
    document.body.classList.remove('overflow-hidden');
  },
  methods: {
    /**
     * Close the modal if the user presses the escape key.
     * @param {Event} e
     */
    handleKeyup(e) {
      if (e.keyCode !== 27) {
        return;
      }
      this.close();
    },
    handleOverlayClick() {
      if (!this.overlayClose) {
        return;
      }
      this.close();
    },
    close() {
      this.$emit('close');
    },
  },
};
</script>

<style scoped>
.modal-dialog__body {
  @apply text-center;
}

.modal-dialog >>> .modal-overlay {
  @apply fixed inset-0 z-40;
}

.modal-dialog >>> .modal-close {
  @apply fixed m-6 right-0 top-0 z-50;
}

.modal-dialog >>> .modal-window {
  @apply fixed left-1/2 top-1/2 transform -translate-x-1/2 -translate-y-1/2 z-50;
}

@media screen and (min-height: 48rem) {
  .modal-dialog >>> .modal-window {
    @apply translate-y-0;
    max-height: calc(100vh - 13rem);
    top: 6.5rem;
  }
}
</style>
