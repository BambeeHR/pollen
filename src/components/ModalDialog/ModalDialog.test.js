import { mount } from '@vue/test-utils';
import ModalDialog from './ModalDialog.vue';

describe('ModalDialog', () => {
  test('renders correctly', () => {
    const wrapper = mount(ModalDialog, {
      slots: {
        default: 'Content goes here',
        header: 'Heading goes here',
        subfooter: 'Subfooter goes here',
      },
      propsData: {
        actions: {
          primary: {
            label: 'Accept',
          },
          secondary: {
            label: 'Skip',
          },
          tertiary: {
            label: 'Delete',
            preIcon: 'trash',
          },
        },
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('emits a close event when escape is pressed', () => {
    const elem = document.createElement('div');
    if (document.body) {
      document.body.appendChild(elem);
    }
    const wrapper = mount(ModalDialog, {
      attachTo: elem,
    });
    wrapper.trigger('keyup.enter');
    expect(wrapper.emitted().close).toBeFalsy();
    wrapper.trigger('keyup.esc');
    expect(wrapper.emitted().close.length).toBe(1);
    wrapper.destroy();
    wrapper.trigger('keyup.esc');
    expect(wrapper.emitted().close.length).toBe(1);
  });
});
