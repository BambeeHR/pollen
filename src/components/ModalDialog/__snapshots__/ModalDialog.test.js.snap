// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalDialog renders correctly 1`] = `
<div
  class="modal-dialog"
>
  <div
    tabindex="0"
  />
   
  <div
    tabindex="1"
  />
   
  <div
    data-lock=""
  >
    <div
      aria-describedby="test-body"
      aria-labelledby="test-header"
      class="modal-window modal-window--regular"
      id="test"
      role="dialog"
    >
      <div
        class="modal-header modal-header--transparent"
        id="test-header"
      >
        Heading goes here
      </div>
       
      <div
        class="modal-dialog__body"
        id="test-body"
      >
        Content goes here
      </div>
       
      <div
        class="modal-actions modal-actions--justify"
      >
        <button
          class="modal-actions__action modal-actions__action--primary base-button base-button--primary base-button--medium base-button--regular base-button--inline"
          type="button"
        >
          <div
            class="base-button__inner"
          >
            Accept
          </div>
        </button>
         
        <button
          class="modal-actions__action modal-actions__action--secondary base-button base-button--secondary base-button--medium base-button--flat base-button--inline"
          type="button"
        >
          <div
            class="base-button__inner"
          >
            Skip
          </div>
        </button>
         
        <button
          class="modal-actions__action modal-actions__action--tertiary base-button base-button--tertiary base-button--medium base-button--flat base-button--inline"
          type="button"
        >
          <div
            class="base-button__inner"
          >
            <svg
              aria-hidden="true"
              class="base-icon svg-inline--fa fa-trash fa-w-14 base-button__icon base-button__icon--pre"
              data-icon="trash"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class=""
                d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                fill="currentColor"
              />
            </svg>
            Delete
          </div>
        </button>
      </div>
       
      <div
        class="modal-subfooter"
      >
        Subfooter goes here
      </div>
    </div>
     
    <button
      class="modal-close modal-close--light"
      title="Close"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="base-icon svg-inline--fa fa-times fa-w-10"
        data-icon="times"
        data-prefix="fal"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M193.94 256L296.5 153.44l21.15-21.15c3.12-3.12 3.12-8.19 0-11.31l-22.63-22.63c-3.12-3.12-8.19-3.12-11.31 0L160 222.06 36.29 98.34c-3.12-3.12-8.19-3.12-11.31 0L2.34 120.97c-3.12 3.12-3.12 8.19 0 11.31L126.06 256 2.34 379.71c-3.12 3.12-3.12 8.19 0 11.31l22.63 22.63c3.12 3.12 8.19 3.12 11.31 0L160 289.94 262.56 392.5l21.15 21.15c3.12 3.12 8.19 3.12 11.31 0l22.63-22.63c3.12-3.12 3.12-8.19 0-11.31L193.94 256z"
          fill="currentColor"
        />
      </svg>
      <div
        aria-hidden="true"
        class="modal-close__visible-label"
      >
        Esc
      </div>
      <div
        class="modal-close__accessible-label"
      >
        Click or press escape to close
      </div>
    </button>
     
    <div
      class="modal-overlay modal-overlay--primary"
    />
  </div>
   
  <div
    tabindex="0"
  />
</div>
`;
